correct_periscope_drift.py

```diff
@@ -0,0 +1,431 @@
+    zag = np.arctan2(d_aca[:, 2], d_aca[:, 0]) * R2A
+    return yag, zag
+
+
+def time_bins(times, x, nbins=20):
+    h, bins = np.histogram(times, bins=nbins)
+    bin_centers = (bins[:-1] + bins[1:]) / 2.0
+    inds = np.digitize(times, bins) - 1
+    bin_x = []
+    bin_std = []
+    for idx in range(0, nbins):
+        data = x[inds == idx]
+
+        bin_x.append(np.mean(data))
+        bin_std.append(np.std(data)/np.sqrt(len(data)))
+    return bin_centers, np.array(bin_x), np.array(bin_std)
+
+
+def fit(fit_data, evt_times, data_id, opt):
+
+    init_error = 5
+
+    ui.clean()
+    ui.load_arrays(data_id, evt_times - evt_times[0], fit_data,
+                   np.zeros_like(fit_data) + init_error)
+    v2("Fitting a line to the data to get reduced stat errors")
+    # First just fit a line to get reduced errors on this set
+    ui.polynom1d.line
+    ui.set_model(data_id, 'line')
+    ui.thaw('line.c1')
+    ui.fit(data_id)
+    fit = ui.get_fit_results()
+    calc_error = init_error * np.sqrt(fit.rstat)
+    ui.set_staterror(data_id, calc_error)
+    # Then fit the specified model
+    v2("Fitting a polynomial of degree {} to the data".format(opt['corr_poly_degree']))
+    ui.polynom1d.fitpoly
+    ui.freeze('fitpoly')
+    # Thaw the coefficients requested by the degree of the desired polynomial
+    for deg in range(1, 1 + opt['corr_poly_degree']):
+        ui.thaw("fitpoly.c{}".format(deg))
+    if opt['corr_poly_degree'] > 1:
+        ui.thaw('fitpoly.offset')
+    ui.set_model(data_id, 'fitpoly')
+    ui.fit(data_id)
+    mp = ui.get_model_plot(data_id)
+    return mp
+
+
+# The '@handle_ciao_errors' decorator will catch any error thrown and
+# display it in a format like that of a CIAO tool, then exit with a
+# non-zero status. So, if at any point in your code you need to exit just
+# raise an error and let handle_ciao_errors bother with the display.
+#
+# This also means that routines can be used from ChIPS/Sherpa/other scripts
+# without having to worry about them calling sys.exit.
+
+
+@handle_ciao_errors(TOOLNAME, VERSION)
+def main(opt):
+    events = extract_events(opt['evtfile'],
+                            opt['x'], opt['y'], opt['radius'])
+
+    evt_ra_nom = events.get_key('RA_NOM').value
+    evt_dec_nom = events.get_key('DEC_NOM').value
+    evt_roll_nom = events.get_key('ROLL_NOM').value
+
+    asol = pycrates.read_file(opt['infile'])
+    asol_times = asol.get_column('time').values
+
+    # Sanity check the two input files
+    asol_obsid = asol.get_key('OBS_ID').value
+    evt_obsid = events.get_key('OBS_ID').value
+    if asol_obsid != evt_obsid:
+        v1("Error Aspect solution obsid {} != event file obsid {}".format(asol_obsid, evt_obsid))
+
+    evt_ra = events.get_column('RA').values
+    evt_dec = events.get_column('Dec').values
+    evt_times = events.get_column('Time').values
+
+    # Limit to only using events contained within the range of the aspect solution
```

Comment:
I don't have an answer for this yet.  I think we presently have discontinuities between aspect solutions that may be greater than the likely size of this correction.  Since this correction is correcting about the mean of events passed to it, doing that correction (with different mean values for each) on multiple files would likely *increase* the discontinuities between those aspect solution.  So I think we should limit this tool to vanilla observations with single aspect intervals and deal with others on a case-by-case.